### ISSUE TITLE
bmz.load("HawkEars").training is True

### ISSUE THREAD

**Issue by lostanlen** [kitzeslab/opensoundscape#1081]
Consider the following script:

```python
from opensoundscape.ml import bioacoustics_model_zoo as bmz
he =  bmz.load("HawkEars")
print(he.training)
```

I am surprised to see that the answer is `True`, which (i think?) has implications in terms of batch normalization, since `track_running_stats` is also true. I believe that the implications is that the distributed model has side effects, in the sense that it is sensitive to the order in which test samples are presented.

What is the rationale behind `bmz.load("HawkEars").training`? Was this intended?

**Comment by sammlapp** [comment]
Hi @lostanlen , I believe the model will work correctly, but I can see how this is confusing for two reasons: (1) the pytorch model that we use for forward/backward is an attribute (`.network`) of the HawkEars object, rather than the object itself; and (2) `.train()` and `.eval()`, as well as enabling and disabling gradients, is done within the class's `train`, `predict`, and `__call__` methods. 

Specifically, HawkEars subclasses the OpenSoundscape SpectrogramClassifier class. The inherited predict() method uses`__call__`, and `__call__` includes
https://github.com/kitzeslab/opensoundscape/blob/d19627825fd2d0f190f5c6d26121973a5849b31f/opensoundscape/ml/cnn.py#L1857-L1858

such that during inference (or the validation phase of training), self.network.training will be False.

That said, your question made me look into this behavior in other places in the code base and it seems we need to correct the behavior of the `generate_cams` method which calls `self.network.forward()` without first setting `self.network.eval()`
https://github.com/kitzeslab/opensoundscape/blob/d19627825fd2d0f190f5c6d26121973a5849b31f/opensoundscape/ml/cnn.py#L2085-L2086

Perhaps we should remove the .training attribute (property?) of the top-level class to avoid confusion (or more generally, if SpectrogramClassifier should not be a child of torch.nn.Module at all). 

Also, if you observe inconsistent inference results in any way (e.g. depending on the order of samples) please do let us know.

**Comment by lostanlen** [comment]
Thank you for your reply. I am not a specialist on batch normalization nor on PyTorch internals. 

I have found this paper: https://arxiv.org/abs/2006.10963

I guess my question is whether this "prediction-time batch normalization" is what OSC implements by default

**Comment by sammlapp** [comment]
Under the hood opensoundscape is always setting `pytorch_model.train()` for inference and `pytorch_model.train()` during training, which is a generally-accepted standard practice. 

The exact effects of this may depend on the implementation of a specific pytorch architecture. In general, though, pytorch models mostly use [BatchNorm1d](https://pytorch.org/docs/main/generated/torch.nn.BatchNorm1d.html#torch.nn.BatchNorm1d), and the behavior of this class is to compute running averages during training that are subsequetly used during inference (not prediction-time batch normalization), as long as track_running_stats=True. (Perhaps track_running_stats=False produces the behavior described as prediction-time batch normalization; we never set track_running_stats=False within the opso code base though). 

As noted [here](https://discuss.pytorch.org/t/why-track-running-stats-is-not-set-to-false-during-eval/25412/2), the batch normalization values are not updated in eval mode, even when track_running_state=True. 

Let me know if I'm missing your point though, I am not an expert on batch normalization either.

**Comment by lostanlen** [comment]
Thanks! this answers my question
